airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/default_stream.py

```diff
@@ -34,7 +36,7 @@ def __init__(
         self._primary_key = primary_key
         self._cursor_field = cursor_field
         self._logger = logger
-        self._cursor = cursor or NoopCursor()
+        self._cursor = cursor or FinalStateCursor(stream_name=name, stream_namespace=namespace, message_repository=message_repository)
```

Comment:
This would be a breaking change (adding `message_repository` already is anyway so this does not seem like an argument in this case) but I'm wondering if it would be more maintainable to have `cursor` not be optional. The consequence of it being optional is that this class needs to know more about the implementation details of the cursor and all its dependencies and its interface.

Reply:
I would tend to think that making it non-optional would make it less maintainable, as now we have to deal with each instantiation of the default from the inheritor (inb4 incorrect terminology but I hope you know what I mean) if we change a smart default. I'm interested in your thoughts on how this is more maintainable though since you are way more familiar with this than i am 😄

Reply:
> as now we have to deal with each instantiation of the default from the inheritor

With both solutions, we will have to deal with this anyway as we now have to pass the message_repository everywhere because the implementation of the default cursor has changed (see the changes made to `airbyte-cdk/python/unit_tests/sources/streams/concurrent/test_default_stream.py` to illustrate that).

In other words, as `default_stream.py` depends on actual implementations of cursors, it is at risk of changes when there is a change in logic in the cursors. Ideally, it should only change when there is a change with respect to the DefaultStream as it is its scope of the domain that it handles.

The value of not having a default is that changes are scoped to their domain. I don't have a strong feeling because it still feels like the debt is small here but I would still prefer having things properly bounded